codex/utils/asynciter.nim

```diff
@@ -49,98 +56,71 @@ proc new*[T](_: type Iter, genNext: GenNext[T], isFinished: IsFinished, finishOn
-    .map((i: int) => items[i])
-
-proc fromSlice*[U, V: Ordinal](_: type Iter, slice: HSlice[U, V]): Iter[U] =
-  ## Creates new iterator from slice
-  ##
-
-  Iter.fromRange(slice.a.int, slice.b.int, 1)
-
-proc fromRange*[U, V, S: Ordinal](_: type Iter, a: U, b: V, step: S = 1): Iter[U] =
-  ## Creates new iterator in range a..b with specified step (default 1)
+proc emptyAsyncIter*[T](): AsyncIter[T] =
+  ## Creates an empty AsyncIter
   ##
 
-  var i = a
+  proc genNext(): Future[T] {.raises: [CatchableError].} =
+    raise newException(CatchableError, "Next item requested from an empty AsyncIter")
+  proc isFinished(): bool = true
 
-  proc genNext(): U =
-    let u = i
-    inc(i, step)
-    u
+  newAsyncIter[T](genNext, isFinished)
 
-  proc isFinished(): bool =
-    (step > 0 and i > b) or
-      (step < 0 and i < b)
-
-  Iter.new(genNext, isFinished)
-
-proc map*[T, U](iter: Iter[T], fn: Function[T, U]): Iter[U] =
-  Iter.new(
-    genNext    = () => fn(iter.next()),
+proc map*[T, U](iter: AsyncIter[T], fn: Function[T, Future[U]]): AsyncIter[U] =
+  newAsyncIter[U](
+    genNext    = () => iter.next().flatMap(fn),
     isFinished = () => iter.finished
   )
 
-proc filter*[T](iter: Iter[T], predicate: Function[T, bool]): Iter[T] =
-  var nextT: Option[T]
+proc mapFilter*[T, U](iter: AsyncIter[T], mapPredicate: Function[T, Future[Option[U]]]): Future[AsyncIter[U]] {.async.} =
+  var nextFutU: Option[Future[U]]
 
-  proc tryFetch(): void =
-    nextT = T.none
+  proc tryFetch(): Future[void] {.async.} =
+    nextFutU = Future[U].none
     while not iter.finished:
-      let t = iter.next()
-      if predicate(t):
-        nextT = some(t)
+      let futT = iter.next()
+      try:
+        if u =? await futT.flatMap(mapPredicate):
+          let futU = newFuture[U]("AsyncIter.mapFilterAsync")
+          futU.complete(u)
+          nextFutU = some(futU)
+          break
+      except CatchableError as err:
+        let errFut = newFuture[U]("AsyncIter.mapFilterAsync")
+        errFut.fail(err)
+        nextFutU = some(errFut)
         break
 
-  proc genNext(): T =
-    let t = nextT.unsafeGet
-    tryFetch()
-    return t
+  proc genNext(): Future[U] {.async.} =
+    let futU = nextFutU.unsafeGet
+    await tryFetch()
+    await futU
 
   proc isFinished(): bool =
-    nextT.isNone
-
-  tryFetch()
-  Iter.new(genNext, isFinished)
+    nextFutU.isNone
 
-proc prefetch*[T](iter: Iter[T], n: Positive): Iter[T] =
-  var ringBuf = newSeq[T](n)
-  var iterLen = int.high
-  var i = 0
+  await tryFetch()
+  newAsyncIter[U](genNext, isFinished)
 
-  proc tryFetch(j: int): void =
-    if not iter.finished:
-      let item = iter.next()
-      ringBuf[j mod n] = item
-      if iter.finished:
-        iterLen = min(j + 1, iterLen)
+proc filter*[T](iter: AsyncIter[T], predicate: Function[T, Future[bool]]): Future[AsyncIter[T]] {.async.} =
+  proc wrappedPredicate(t: T): Future[Option[T]] {.async.} =
+    if await predicate(t):
```

Comment:
Why do we care about handling errors when doing mapFilter but not when doing filter?

Reply:
Had missed that.